File: photonflow-cpp/src/blocks/optical/coupler.cpp
/**
 * @file coupler.cpp
 * @brief 2x2 Optical coupler block implementation.
 */

#include "photonflow/blocks/optical/coupler.hpp"
#include "photonflow/blocks/block_registry.hpp"

#include <algorithm>
#include <cmath>
#include <complex>


namespace photonflow {

Coupler::Coupler(const std::string &id, const json &params,
                 const json &nonideal)
    : BaseBlock(id, params, nonideal) {}

std::unordered_map<std::string, Signal>
Coupler::process(const std::unordered_map<std::string, Signal> &inputs,
                 SimContext & /*ctx*/) {
  const Signal &in1 = inputs.at("opt_in1");

  // Get second input or create zero signal
  Eigen::VectorXcd in2_data;
  if (inputs.contains("opt_in2")) {
    in2_data = inputs.at("opt_in2").data;
  } else {
    in2_data = Eigen::VectorXcd::Zero(in1.n_samples());
  }

  // Get splitting ratio and clamp to [0, 1]
  double k = get_param("split_ratio", 0.5);
  k = std::clamp(k, 0.0, 1.0);

  // Non-ideal effects
  bool nonideal_enable = get_nonideal("enable", false);
  double phi_err = 0.0;
  double loss_db = 0.0;

  if (nonideal_enable) {
    double delta_k = get_nonideal("split_ratio_error", 0.0);
    k = std::clamp(k + delta_k, 0.0, 1.0);
    phi_err = get_nonideal("phase_error", 0.0);
    loss_db = get_nonideal("loss_db", 0.0);
  }

  // Coupler matrix coefficients
  // [a  b]   [  sqrt(k)           j*sqrt(1-k)*exp(j*phi_err) ]
  // [c  d] = [ j*sqrt(1-k)        sqrt(k)                     ]
  double a = std::sqrt(k);
  std::complex<double> b = std::complex<double>(0.0, 1.0) * std::sqrt(1.0 - k) *
                           std::exp(std::complex<double>(0.0, phi_err));
  std::complex<double> c = std::complex<double>(0.0, 1.0) * std::sqrt(1.0 - k);
  double d = std::sqrt(k);

  const Eigen::Index n = in1.n_samples();
  Eigen::VectorXcd out1_data(n);
  Eigen::VectorXcd out2_data(n);

  for (Eigen::Index i = 0; i < n; ++i) {
    out1_data[i] = a * in1.data[i] + b * in2_data[i];
    out2_data[i] = c * in1.data[i] + d * in2_data[i];
  }

  // Apply loss
  if (nonideal_enable && loss_db != 0.0) {
    double loss_linear = std::pow(10.0, -loss_db / 20.0);
    out1_data *= loss_linear;
    out2_data *= loss_linear;
  }

  Signal sig1(out1_data, in1.fs, in1.t0);
  sig1.center_freq = in1.center_freq;
  sig1.pol_mode = in1.pol_mode;
  sig1.meta = in1.meta;

  Signal sig2(out2_data, in1.fs, in1.t0);
  sig2.center_freq = in1.center_freq;
  sig2.pol_mode = in1.pol_mode;
  sig2.meta = in1.meta;

  return {{"opt_out1", std::move(sig1)}, {"opt_out2", std::move(sig2)}};
}

std::optional<std::string> Coupler::port_type(const std::string &port) const {
  if (port == "opt_in1" || port == "opt_in2" || port == "opt_out1" ||
      port == "opt_out2") {
    return "optical";
  }
  return std::nullopt;
}

json Coupler::describe() const {
  return {{"ports",
           {{"opt_in1", "optical"},
            {"opt_in2", "optical"},
            {"opt_out1", "optical"},
            {"opt_out2", "optical"}}},
          {"spec",
           {{"params",
             {{"split_ratio",
               {{"type", "float"}, {"default", 0.5}, {"unit", ""}}}}},
            {"nonideal",
             {{"enable", {{"type", "bool"}, {"default", false}}},
              {"split_ratio_error",
               {{"type", "float"}, {"default", 0.0}, {"unit", ""}}},
              {"phase_error",
               {{"type", "float"}, {"default", 0.0}, {"unit", "rad"}}},
              {"loss_db",
               {{"type", "float"}, {"default", 0.0}, {"unit", "dB"}}}}}}}};
}

REGISTER_BLOCK(Coupler, "Coupler");

} // namespace photonflow

File: backend/src/photonflow/blocks/optical/coupler.py
"""2x2 optical coupler."""

from __future__ import annotations

import math
from typing import Dict

import torch

from photonflow.blocks.base import BaseBlock, register_block
from photonflow.core.signal import Signal
from photonflow.core.sim import SimContext


@register_block("Coupler")
class Coupler(BaseBlock):
    """2×2 optical coupler/splitter. Combines or splits optical signals with configurable coupling ratio and phase."""

    PORTS = {
        "opt_in1": "optical",
        "opt_in2": "optical",
        "opt_out1": "optical",
        "opt_out2": "optical",
    }
    SPEC = {
        "params": {
            "split_ratio": {"type": "float", "default": 0.5, "unit": ""},
        },
        "nonideal": {
            "enable": {"type": "bool", "default": False},
            "split_ratio_error": {"type": "float", "default": 0.0, "unit": ""},
            "phase_error": {"type": "float", "default": 0.0, "unit": "rad"},
            "loss_db": {"type": "float", "default": 0.0, "unit": "dB"},
        },
    }

    def process(self, inputs: Dict[str, Signal], ctx: SimContext) -> Dict[str, Signal]:
        in1 = inputs["opt_in1"]
        in2 = inputs.get("opt_in2")
        if in2 is None:
            in2 = Signal(
                data=torch.zeros_like(in1.data),
                fs=in1.fs,
                t0=in1.t0,
                center_freq=in1.center_freq,
                pol_mode=in1.pol_mode,
            )

        k = float(self.params.get("split_ratio", 0.5))
        k = min(max(k, 0.0), 1.0)
        nonideal = self.nonideal if self.nonideal.get("enable", False) else {}
        delta_k = float(nonideal.get("split_ratio_error", 0.0))
        k = min(max(k + delta_k, 0.0), 1.0)
        phi_err = float(nonideal.get("phase_error", 0.0))

        a = math.sqrt(k)
        b = 1j * math.sqrt(1.0 - k) * complex(math.cos(phi_err), math.sin(phi_err))
        c = 1j * math.sqrt(1.0 - k)
        d = math.sqrt(k)

        out1 = a * in1.data + b * in2.data
        out2 = c * in1.data + d * in2.data

        loss_db = nonideal.get("loss_db")
        if loss_db is not None:
            loss = 10 ** (-float(loss_db) / 20.0)
            out1 = out1 * loss
            out2 = out2 * loss

        sig1 = Signal(
            data=out1,
            fs=in1.fs,
            t0=in1.t0,
            center_freq=in1.center_freq,
            pol_mode=in1.pol_mode,
            meta=dict(in1.meta),
        )
        sig2 = Signal(
            data=out2,
            fs=in1.fs,
            t0=in1.t0,
            center_freq=in1.center_freq,
            pol_mode=in1.pol_mode,
            meta=dict(in1.meta),
        )
        return {"opt_out1": sig1, "opt_out2": sig2}
